### ISSUE TITLE
Crash when using landscape rotations

### ISSUE THREAD

**Issue by marcoms** [baskerville/plato#129]
Whenever I use Plato in landscape more for a while it crashes/freezes and the device requires a restart. Seems to be linked to when I tap the page/use a gesture. I don't see a crash.log anywhere in the filesystem (does it need to be a debug build?) but here is an info.log:

```
Plato is running on a Kobo Clara HD.
The framebuffer resolution is 1072 by 1448.
Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.cbz
Add new entry: 4C3E472C02B3972B, Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.cbz.
Plato is running on a Kobo Clara HD.
The framebuffer resolution is 1072 by 1448.
Remove entry: 4C3E4314012A8825, Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.epub.
Remove entry: 4C3E419202DB3A93, Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.pdf.
Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.cbz
Framebuffer rotation: 0 -> 0.
Add new entry: 4C3E472C02B3972B, Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.cbz.
Plato is running on a Kobo Clara HD.
The framebuffer resolution is 1072 by 1448.
Went to sleep on July 14, 2020 at 01:55.
Woke up on July 14, 2020 at 01:58.
udhcpc: started, v1.31.1.kobo
udhcpc: sending discover
udhcpc: received SIGTERM
Plato is running on a Kobo Clara HD.
The framebuffer resolution is 1072 by 1448.
Hunter x Hunter, Vol. 33 - Yoshihiro Togashi.cbz
Framebuffer rotation: 2 -> 2.
Framebuffer rotation: 1 -> 1.
Framebuffer rotation: 0 -> 0.
```

**Comment by marcoms** [comment]
Apologies for submitting too early, edited with the correct description now

**Comment by baskerville** [comment]
If you have to reboot manually, it means *Plato* hasn't crashed (and therefore `info.log` doesn't contain any clue).

Knowing whether this is a software or hardware bug would be useful.

In order to known that, you'd have to access your device through Telnet and attach to the running *Plato* process with GDB (you'll need a release build with debug symbols).

You should find the GDB binaries in @NiLuJe's [Kobo Stuff](https://www.mobileread.com/forums/showthread.php?t=254214) package.

**Comment by NiLuJe** [comment]
You appear to be reading a CBZ.

Did it hang right after a page-turn?

There are still random crash bugs left in Mk. 7 kernels, and one of the "easiest" ways to trigger one of 'em seem to be displaying images, a lot of images ;p.

I never managed to reproduce one w/ SSH running, so I don't know if it's a hard crash, or simply the EPDC giving up the ghost (like what happens on older, broken devices when you feed 'em a broken mxcfb_rect).

TL;DR: Yeah, more info needed, but Plato might not have anything to do w/ it.

**Comment by marcoms** [comment]
@baskerville Thanks for clearing that up, I've tried to run gdb from the package you linked but it seems to not work due to the `readline` library being missing. I've reported this on the forum post but it could be that I've done something wrong.

To compile a release build with debug symbols, would I need to do something like this?

```
RUSTFLAGS="-g" ./build.sh
```

**Comment by marcoms** [comment]
@NiLuJe IIRC it even happened when reading an ePub with mostly text, also rotated to landscape (0 degrees). I think turning the page did trigger it but it also happened when opening/closing the menu

**Comment by NiLuJe** [comment]
@marcoms:

Gah. I'll fix the gdb issue in the next build. If you want a quick hotfix, the missing readline lib ships as part of (either of) the Python package (... and those are in gdb's rpath, which is probably why I missed it, as I usually have Py3 installed ^^).

----

I've also seen in happen on text, FWIW, but much, much, much, much rarely.

Rota 0 (as in, the actual linuxfb rotate field set to 0, what it actually corresponds to in practice is device-specific) may be special-cased in some codepaths on some devices (I don't recall to which extent that holds true on the Clara), so this may be a compounding factor here.

**Comment by marcoms** [comment]
Thanks for responding so soon. I did some further testing, trying for ages trying to reproduce when on 180 but I wasn't able to. After switching to 0 it froze again after a few seconds of scrolling... Looks like there is something fishy with 0 🤔

**Comment by baskerville** [comment]
> To compile a release build with debug symbols, would I need to do something like this?
> 
> ```
> RUSTFLAGS="-g" ./build.sh
> ```

Yes, and the resulting binary will be `target/arm-unknown-linux-gnueabihf/release/plato`.

**Comment by NiLuJe** [comment]
As I mentioned earlier, rota 0 is a compounding factor, and rota 0 + 8bpp even more so, so, my guess would be since the switch to 8bpp ;).

**Comment by rien333** [comment]
I got gdb setup, and have plato compiled with debug symbols. I'm not an expert on gdb (or in other domains related to this problem, probably), so I might need a little guidance. 

I attached to plato's PID, then ran `continue` in gdb, triggered the crash/freeze in plato, but nothing really interesting happened in gdb. The only messages I saw were of the form 

```
[New LWP 1234]
[LWP 1234 exited]
```

After the crash/freeze, there was no further output in gdb. plato still reacted to touch events, however, and everytime I touched the screen it shows something like:

```
[New LWP 5493]
[New LWP 5494]
[New LWP 5495]
[LWP 5495 exited]
[LWP 5493 exited]
[LWP 5494 exited]
```

Anything else I can do/should be doing differently?

**Comment by NiLuJe** [comment]
I'd be glad for a look at `dmesg` when this happens, because I still haven't been able to reproduce one of those with a shell attached :s.

But given that your shell was still alive, this very much looks like the EPDC crashing and burning, like the usual EPDC issues of the sort (e.g., 1x1 refreshes on some devices, or other EPDC hangs).

**Comment by NiLuJe** [comment]
I'd add a request for an ioctl strace log, but, IIRC, the ioctl generally return successfully in most of these cases. And I *think* Plato itself should be handling ioctl failures sanely, anyway, so that would have been logged.

TL;DR: You probably don't have to bother with strace ;p.

**Comment by baskerville** [comment]
@rien333 To check @NiLuJe insight, remove all the lines that contain *fbdepth* from `plato.sh`.

**Comment by rien333** [comment]
> I'd be glad for a look at dmesg when this happens, because I still haven't been able to reproduce one of those with a shell attached :s.

Nothing interesting, I think. dmesg always floods with RTL871X related messages (which relates to wifi I think), so I cut some out. 

I'll try your guys'  other suggestions in a bit.

```
RTL871X: UpdateHalRAMask8188F => mac_id:0, networkType:0x0b, mask:0x000fffff
	==> rssi_level:1, rate_bitmap:0x000f0000
RTL871X: rtl8188f_set_FwMacIdConfig_cmd(): mac_id=0 raid=0x3 bw=0 mask=0xf0000
...
RTL871X: rtl8188f_hal_dm_in_lps, RSSI_Min=56
PMU: mainFlowOfLowVoltage :	Noxx : cc_delta_cap is -9, cc_delta_cap_now is -24, last_cc_delta_cap is -15
PMU: mainFlowOfLowVoltage :	Noxx : temp_cc_delta_cap is 0, after temp_cc_delta_cap_mas is -159, cc_cap_mas -12235
PMU: mainFlowOfLowVoltage :	No9 :Cap is 7290 , low_rate is 68, dsoc is 6819, capnow is 7240, capzero is 0, delta cc is -6, delta cc ori is -9
PMU: mainFlowOfLowVoltage :	No10 :temp_mas is -218, offset_mas is -59, value is -612, final value is -6
PMU:STATUS= 7: IBAT= -161: VSYS= 3892250: VBAT= 3903800: DSOC= 6819: RSOC= 6800: cc_delta=77: rrf= 0
RTL871X: CurIGValue=56, EntryMinUndecoratedSmoothedPWDB = 59
tps6518x_get_temperature():temperature = 30
```

**Comment by rien333** [comment]
> @rien333 To check @NiLuJe's insight, remove all the lines that contain fbdepth from plato.sh.

This seems to avoid the freeze. Let me know if there's anything else I can do!

**Comment by baskerville** [comment]
> > @rien333 To check @NiLuJe's insight, remove all the lines that contain fbdepth from plato.sh.
> 
> This seems to avoid the freeze.

Thanks, I'll fix `plato.sh`.

@NiLuJe Are all the mark 7 devices concerned?

**Comment by rien333** [comment]
I have the same kobo as OP (Kobo Clara HD, a mark 7 device), btw.

**Comment by NiLuJe** [comment]
@baskerville: The special snowflake aspect of rota 0 + bpp8 exists since day one of the i.MX EPDCs, but it would appear to only be extra-screwy on EPDCv2 devices with NTX kernels, so, yeah, every Mk. 7.
